File: assn7/main.cpp
#include <cstdio>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <utility>
#include <vector>
using namespace std;

/* Input file notes:
 * DMFB - "arch.in"
 *      - M and N = DMFB dimensions
 *      - N_input and N_output = number of in/out reservoirs
 *      - N_sensor, N_detector, N_heater 
 *          = number of integrated sensors/detectors/heaters
 *      - N_input lines, ea. line = (x, y) pair for in reservoir location
 *      - N_output lines, ea. line = (x, y) pair for out reservoir location
 *      - N_sensor lines, ea. line = (x, y) pair for sensor locations
 *      - N_detector lines, ea. line = (x, y) pair for detector locations
 *      - N_heater lines, ea. line = (x, y) pair for heater locations
 *      - ASSUME: i/o reservoirs are always on perimeter
 *                all device coords are within rows 2 to M-1, columns 2 to N-1
 * Assay Ops - "ops.in"
 *      - N_ops = number of ops in assay
 *      - N_ops lines, ea. line describes an assay op w/ id #s 1-...-N_ops
 *      - input format: 1 r_in
 *          - r_in = positive integer ID of input reservoir
 *      - output format: 2 r_out
 *          - r_out = positive integer ID of output reservoir
 *      - mixing format: 3 z(v)
 *          - z(v) = cost value, represents importance of mix (FLOAT VALUE)
 *      - other formats: t(v) 
 *          - int type: split = 4, merge = 5, store = 6, sense = 7, detect = 8
 *                      heat = 9
 * Graph file - "graphs.in"
 *      - N_ops, e_int, e_com 
 *          - N_ops = number of assay ops, number of vertices
 *          - e_int = number of edges in interference graph
 *          - e_com = number of edges in communication graph
 *      - e_int lines, enumerate edge set of interference graph
 *          - form: u v
 *      - e_com lines, enumerate edge set of communication graph
 *          - form: u v c(f) 
 *
 * Alpha Parameter File - "alpha.in"
 *      - single floating point value for alpha
 */ 

void printCoords(vector<pair<int,int> > x) {
    for(unsigned i = 0; i < x.size(); i++) {
        cout << "(" << x.at(i).first << ", " << x.at(i).second << ")" << endl;
    }
    return; 
}

void printVec(vector<vector<int> > x) {
    for(unsigned i = 0; i < x.size(); i++) {
        for (unsigned j = 0; j < x.at(i).size(); j++) { 
            cout << x.at(i).at(j) << " ";
        }
        cout << endl;
    }
    return;
}

int main (int argc, char **argv) {

    ifstream inFS;

    //temp vars for pair coord creation
    int tempX = 0;
    int tempY = 0;

    /****************************READ ARCH.IN****************************/
    //DMFB variables
    int dim_M = 0;
    int dim_N = 0;
    int n_input = 0;
    int n_output = 0; 
    int n_sensor = 0;
    int n_detector = 0;
    int n_heater = 0;

    printf("Reading from arch.in... \n");
    inFS.open("arch.in");
    if (!inFS.is_open()) {
        perror("arch.in open failed");
        return EXIT_FAILURE;
    }
    inFS >> dim_M >> dim_N >> n_input >> n_output 
        >> n_sensor >> n_detector >> n_heater;

    cout << "dim_M = " << dim_M << endl;
    cout << "dim_N = " << dim_N << endl;
    cout << "n_input = " << n_input << endl;
    cout << "n_output = " << n_output << endl; 
    cout << "n_sensor = " << n_sensor << endl;
    cout << "n_detector = " << n_detector << endl;
    cout << "n_heater = " << n_heater << endl;

    //input coordinates
    vector<pair<int,int> > inputCoords;
    for(unsigned i = 0; i < n_input; i++) {
        inFS >> tempX >> tempY;
        inputCoords.push_back(make_pair(tempX, tempY));
    }
    cout << "Input Coords: " << endl;
    printCoords(inputCoords);

    //output coordinates
    vector<pair<int,int> > outputCoords;
    for(unsigned i = 0; i < n_input; i++) {
        inFS >> tempX >> tempY;
        outputCoords.push_back(make_pair(tempX, tempY));
    }
    cout << "Output Coords: " << endl;
    printCoords(outputCoords);

    //sensor coordinates
    vector<pair<int,int> > sensorCoords;
    for(unsigned i = 0; i < n_input; i++) {
        inFS >> tempX >> tempY;
        sensorCoords.push_back(make_pair(tempX, tempY));
    }
    cout << "Sensor Coords: " << endl;
    printCoords(sensorCoords);

    //detector coordinates
    vector<pair<int,int> > detectorCoords;
    for(unsigned i = 0; i < n_input; i++) {
        inFS >> tempX >> tempY;
        detectorCoords.push_back(make_pair(tempX, tempY));
    }
    cout << "Detector Coords: " << endl;
    printCoords(detectorCoords);
    
    //heater coordinates
    vector<pair<int,int> > heaterCoords;
    for(unsigned i = 0; i < n_input; i++) {
        inFS >> tempX >> tempY;
        heaterCoords.push_back(make_pair(tempX, tempY));
    }
    cout << "Heater Coords: " << endl;
    printCoords(heaterCoords);
    inFS.close();

    /****************************READ OPS.IN****************************/
    //Assay Ops Variables
    int n_assay_ops = 0;
    float mix_importance = 0.0; 
    vector<int> ops;

    printf("\nReading from ops.in... \n");
    cout << "TODO: COMPLETE ASSAY OPS" << endl;
    inFS.open("ops.in");
    if (!inFS.is_open()) {
        perror("ops.in open failed");
        return EXIT_FAILURE;
    }
    inFS >> n_assay_ops;
    cout << "n_assay_ops = " << n_assay_ops << endl;
    for(unsigned i = 0; i < n_assay_ops; i++) {
    }
 /* Assay Ops - "ops.in"
 *      - N_ops = number of ops in assay
 *      - N_ops lines, ea. line describes an assay op w/ id #s 1-...-N_ops
 *      - input format: 1 r_in
 *          - r_in = positive integer ID of input reservoir
 *      - output format: 2 r_out
 *          - r_out = positive integer ID of output reservoir
 *      - mixing format: 3 z(v)
 *          - z(v) = cost value, represents importance of mix (FLOAT VALUE)
 *      - other formats: t(v) 
 *          - int type: split = 4, merge = 5, store = 6, sense = 7, detect = 8
 *                      heat = 9
 */
    //fill vars
    inFS.close(); 

    /****************************READ GRAPHS.IN****************************/
    //Graph Variables
    int n_ops = 0;
    int e_int = 0;
    int e_com = 0;
    printf("\nReading from graphs.in... \n");
    inFS.open("graphs.in");
    if (!inFS.is_open()) {
        perror("graphs.in open failed");
        return EXIT_FAILURE;
    }
    inFS >> n_ops >> e_int >> e_com;

    //generate matrix representation of interference graph
    cout << "\nGenerating Interference Graph: " << endl;
    vector<vector<int> > interGraph(n_ops, vector<int> (n_ops));
    for(unsigned i = 0; i < e_int; i++) {
        inFS >> tempX >> tempY;
        cout << "Edge from " << tempX << " to " << tempY << endl;
        interGraph.at(tempX-1).at(tempY-1) = 1;
        interGraph.at(tempY-1).at(tempX-1) = 1;
    }
    printVec(interGraph);

    //generate matrix representation of communication graph
    cout << "\nGenerating Communication Graph: " << endl;
    vector<vector<int> > commGraph(n_ops, vector<int> (n_ops));
    for(unsigned i = 0; i < e_com; i++) { 
        inFS >> tempX >> tempY;
        cout << "Edge from " << tempX << " to " << tempY << endl;
        commGraph.at(tempX-1).at(tempY-1) = 1;
        commGraph.at(tempY-1).at(tempX-1) = 1;
    }
    printVec(commGraph);

    inFS.close();

    float alpha = 0.0;
    /****************************READ ALPHA.IN****************************/
    printf("\nReading from alpha.in... \n");
    inFS.open("alpha.in");
    if (!inFS.is_open()) {
        perror("alpha.in open failed");
        return EXIT_FAILURE;
    }
    inFS >> alpha; 
    cout << "alpha = " << alpha << endl;
    inFS.close();

    /****************************BEGIN PLACEMENT****************************/
    vector<vector<int> > solGrid(dim_M, vector<int> (dim_N));

    return 0;
}

File: assn7/main.py
from platypus import NSGAII, DTLZ2
import os
import sys


def read_file(filename):
    print("Reading from " + filename + "...")
    raw_in = []
    f = open(filename, "r")
    for line in f:
        raw_in.append(line.split())
    f.close()
    return raw_in

def write_line(x, f):
    for item in x:
        f.write(str(item) + " ")
    return

def write_results(legal = False, placement = []):
    #TODO: build output function
    """ output file format:
        - Obj(P,C,L)  D_comm(P,C)  T_mix(P,L)  alpha
            - value of objective function (pos float)
            - total communication distance (pos int)
            - latency improvement (pos float)
            - user-specified param (pos float)

        - n_ops lines describing placement location of each operation
            - input operation: 1  r_in
            - output operation: 2  r_out
            - assay operation (mix): 3  x  y  m  n
                - (x, y) grid coordinate of upper left hand position of mixer
                - (m x n) mixer size
            - other assay operations: t(v)  x  y
                - t(v) pos int type for operation  
                - (x, y) grid coordinate of upper left hand position of mixer
    """
    f = open("placement.out", "w")
    if not legal:
        f.write("No solution found")
    else:
        for l in placement:
            write_line(l, f)
            f.write("\n")
    f.close()

def int_map(x):
    y = []
    for i in range(len(x)):
        y.append(map(int, x[i]))
    return y

def print_list(x):
    for l in x:
        print l
    return


def main(debug = "none"):

    # ARCH.IN - DMFB Variables
    dim_M = 0
    dim_N = 0
    n_input = 0
    n_output = 0
    n_sensor = 0
    n_detector = 0
    n_heater = 0
    inputCoords = []
    outputCoords = []
    sensorCoords = []
    detectorCoords = []
    heaterCoords = []

    ################### READ AND SPLIT ARCH.IN ###################
    arch_in = int_map(read_file("arch.in"))
    dim_M = arch_in[0][0]
    dim_N = arch_in[0][1]
    n_input = arch_in[1][0]
    n_output = arch_in[1][1]
    n_sensor = arch_in[2][0]
    n_detector = arch_in[2][1]
    n_heater = arch_in[2][2]
    arch_in = arch_in[3:]
    if n_input > 0:
        inputCoords = arch_in[:n_input]
        arch_in = arch_in[n_input:]
    if n_output > 0:
        outputCoords = arch_in[:n_output]
        arch_in = arch_in[n_output:]
    if n_sensor > 0:
        sensorCoords = arch_in[:n_sensor]
        arch_in = arch_in[n_sensor:]
    if n_detector > 0:
        detectorCoords = arch_in[:n_detector]
        arch_in = arch_in[n_detector:]
    if n_heater > 0:
        heaterCoords = arch_in[:n_heater]
        arch_in = arch_in[n_heater:]
    del arch_in

    # print contents for debugging purposes
    if debug == "arch" or debug == "all":
        print("\nARCH CONTENTS") 
        print("dim_M = " + str(dim_M))
        print("dim_N = " + str(dim_N))
        print("n_input = " + str(n_input))
        print("n_output = " + str(n_output))
        print("n_sensor = " + str(n_sensor))
        print("n_detector = " + str(n_detector))
        print("n_heater = "  + str(n_heater))
        print(inputCoords)
        print(outputCoords)
        print(sensorCoords)
        print(detectorCoords)
        print(heaterCoords)
        print("\n") 

    ################### READ AND SPLIT OPS.IN ###################
    n_assay_ops = 0
    ops_in = read_file("ops.in")
    n_assay_ops = int(ops_in[0][0])
    ops_in = ops_in[1:]
    ops_list = []
    for row in ops_in:
        if row[0] == "3":
            ops_list.append([int(row[0]), float(row[1])])
        else:
            ops_list.append(map(int, row))
    # print contents for debugging purposes
    if debug == "ops" or debug == "all":
        print("n_assay_ops = " + str(n_assay_ops))
        print("ops_list: ")
        print_list(ops_list)

    ################### READ AND SPLIT GRAPHS.IN ###################
    n_ops = 0
    e_int = 0
    e_com = 0
    graphs_in = int_map(read_file("graphs.in"))
    n_ops = graphs_in[0][0]
    e_int = graphs_in[0][1]
    e_com = graphs_in[0][2]
    graphs_in = graphs_in[1:]
    inter_graph = [[0 for x in range(n_ops)] for y in range(n_ops)]
    comm_graph = [[0 for x in range(n_ops)] for y in range(n_ops)]

    # generate interference graph
    if debug == "graphs" or debug == "all":
        print("Generating Interference Graph...")
    for i in range(e_int): 
        # Test output to see which edges were found per graph
        if debug == "graphs" or debug == "all":
            print("Edge from " + str(graphs_in[i][0]) + " to " + str(graphs_in[i][1]))
        tempX = graphs_in[i][0]-1
        tempY = graphs_in[i][1]-1
        inter_graph[tempX][tempY] = 1
        inter_graph[tempY][tempX] = 1
    graphs_in = graphs_in[e_int:]

    # generate communication graph
    if debug == "graphs" or debug == "all":
        print("Generating Communication Graph...")
    for i in range(e_com): 
        if debug == "graphs" or debug == "all":
            print("Edge from " + str(graphs_in[i][0]) + " to " + str(graphs_in[i][1]))
        tempX = graphs_in[i][0]-1
        tempY = graphs_in[i][1]-1
        comm_graph[tempX][tempY] = 1
        comm_graph[tempY][tempX] = 1
    del graphs_in

    # print contents for debugging purposes
    if debug == "graphs" or debug == "all":
        print("\nGRAPHS CONTENTS") 
        print("n_ops = " + str(n_ops))
        print("e_int = " + str(e_int))
        print("e_com = " + str(e_com))
        print("Interference Graph: ")
        print_list(inter_graph)
        print("Communication Graph: ") 
        print_list(comm_graph)
        print("\n")
    
    ################### READ ALPHA.IN ###################
    alpha = float(read_file("alpha.in")[0][0])
    # print contents for debugging purposes
    if debug == "alpha" or debug == "all":
        print("alpha = " + str(alpha))
        print("\n")

    #TODO: Define NSGAII problem given contraints from input files
    ################### NSGA PLACER ###################
    """ Notes:
        - must determine placement of all operations
        - must determine optimal dimensions of each mixing operation
            - mixing operation has sizes of 2x2, 2x3(3x2), 1x4(4x1), 2x4(4x2)
        - start with default NSGAII parameters
        - if a communication link exists between u and v, reduce communication
            cost by placing v near u

        *** Characteristics of a legal solution
        - All assay operations placed within (M-2)x(N-2) subgrid
            -I/O operations are pre-placed on perimeter, cannot be undone
        - For interference edge e = (u,v) in E, u and v cannot overlap
        - Each sensing operation must be placed on electrode with a sensor
        - Each detection operation must be placed on electrode with a detector
        - Each heating operation must be placed on electrode with a heater 
    """


    legal = False # flag to determine if legal solution can be found
    placement = [] # list containing final solution  
    
    
    ################### WRITE RESULTS OUT ###################

    write_results(legal, placement)

    return

if __name__ == "__main__":
    # params to assist with debugging
    if len(sys.argv) > 1 and sys.argv[1].lower() == "debug":
        if len(sys.argv) == 3:
            main(debug = sys.argv[2].lower())
        else:
            main(debug = "all")
    else:
        main()
